Delete expected version

GrantPSpencer opened the pull request:
### Issues

- [x] My PR addresses the following Helix issues and references them in the PR description:

N/A

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
ZooKeeper allows for calling delete with an expected version, where the znode will only be deleted if the version in the znode's stat matches the passed version number. 

This PR adds APIs to the ZKClients to call delete with expected version
Also adds to the baseDataAccessors the API for remove with expected version. 

### Tests

- [x] The following tests are written for this issue:
```
TestZKBaseDataAccessor.java
- testRemoveWithExpectedVersion

RealmAwareZkClientFactoryTestBase.java
- testDeleteExpectedVersion


TestFederatedZkClient.java
- testDeleteExpectedVersion
```

- [x] The following is the result of the "mvn test" command on the appropriate module:
CI run against my fork passed
https://github.com/GrantPSpencer/helix/actions/runs/7909844995/job/21591541741?pr=5

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

N/A. The original delete and remove methods just call the new method and pass in -1 as the expected version (meaning no version match is forced) 

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

desaikomal commented:
Any particular reason for adding this API ?

GrantPSpencer commented:
> Any particular reason for adding this API ?

@desaikomal For upcoming functionality where multiple actors can put cluster into maintenance mode for different reasons. Example, if cluster is maintenance mode for reason_A, reason_B, and reason_C, the cluster will not fully exit and delete the node until all reasons have been removed. ZKHelixAdmin will have API to update the cluster maintenance signal and it will also delete the znode if there are no more reasons left in it. 

However, there is a possible race condition where a reason was added after the client retrieves the znode information but before delete is called. This can be avoided by requiring the znode versions to match when we delete.

junkaixue commented on helix-core/src/main/java/org/apache/helix/manager/zk/ZkBaseDataAccessor.java:
```diff
@@ -720,11 +720,24 @@ public Stat getStat(String path, int options) {
       // despite real error, operation will throw exception when path not empty, and in this
       // case, we try to delete recursively
-      _zkClient.delete(path);
+      _zkClient.delete(path, expectedVersion);
```
NIT: add comment about -1 behavior.

junkaixue reviewed (commented):
The implementation looks fine to me. Be careful with when to use it. When it failed to delete, you need to handle it properly as the exception has been swallowed in the inner part.

GrantPSpencer commented:
> The implementation looks fine to me. Be careful with when to use it. When it failed to delete, you need to handle it properly as the exception has been swallowed in the inner part.

Do you think this should throw ZKBadversion exception at the dataAccessor.remove() level as well? Catching and throwing false is in line with current behavior of remove() method, but it is not clear for what reason the removal failed. Maybe we should throw it so that callers can implement logic specific to version mismatches?

junkaixue commented:
> > The implementation looks fine to me. Be careful with when to use it. When it failed to delete, you need to handle it properly as the exception has been swallowed in the inner part.
> 
> Do you think this should throw ZKBadversion exception at the dataAccessor.remove() level as well? Catching and throwing false is in line with current behavior of remove() method, but it is not clear for what reason the removal failed. Maybe we should throw it so that callers can implement logic specific to version mismatches?

That's what I was worrying about. The failure of deleting znode with version today was not a behavior in Helix.

MarkGaox commented on zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/IZkConnection.java:
```diff
@@ -44,6 +44,7 @@ public interface IZkConnection {
     public String create(String path, byte[] data, List<ACL> acl, CreateMode mode, long ttl) throws KeeperException, InterruptedException;
 
     public void delete(String path) throws InterruptedException, KeeperException;
+    public void delete (String path, int expectedVersion) throws InterruptedException, KeeperException;
```
nit: empty line between two methods.

junkaixue commented:
Anyway, I am totally fine with current change.

GrantPSpencer commented:
@junkaixue @MarkGaox @xyuanlu 
Thanks for reviews. Unfortunately commit history got messed up and had to force push to fastforward. I addressed comments in the most recent commit  (["remove with expected version no longer recursively deletes"](https://github.com/apache/helix/pull/2759/commits/8b9b5d09e5ee412eed53653e1c7f14d0efdbe312))

Important change is that I've decoupled remove and "remove with expected version" in the BaseDataAcessor. Original remove method attempts to recursively delete if the in initial delete fails. The new "remove with expected version" method should not have this recursive delete behavior as it can lead to race condition between initial delete and call to recursive delete. It will now fail if the property is not a leaf node. 
Should I now rename remove with expected version to a separate method entirely or stay with the method overloading? The difference in recursive behavior may not be immediately clear to customers. I appreciate your guys' help in improving my PR

junkaixue commented:
Then I would suggest to have a separate API like deleteWithExpectedVersion. Make it clear it just work for single layer not recursively.

junkaixue commented:
Any update?

GrantPSpencer commented:
@junkaixue sorry was on vacation. This PR is ready for final review. I've changed the naming of the methods to removeWithExpectedVersion as it does not attempt to recursively delete the node like the original remove method does

junkaixue reviewed (approved):
lgtm

GrantPSpencer commented:
Pull request approved by @junkaixue 
Commit message: Add delete with expected version API to zk client and data accessor
